### ISSUE TITLE
Make DatabaseManager Usable Anywhere

### ISSUE THREAD

**Issue by bastisawesome** [bastisawesome/guessinggame_ttv#54]
Title may change, I'm not good at titles...

The current implementation of the DatabaseManager class expects the instances to be wrapped in a context manager and then passed around. While this is fine, and in my opinion ideal as it guarantees the database connection will be closed, it adds a lot of complexity to GGTTV. Notably, a single instance must be maintained for the lifetime of the application, and there are no protections for accidentally creating a new instance. A new instance would, at best, cause an unhandled exception, or at worst, cause race conditions and data corruption.

My proposal is to make the DatabaseManager a static class that maintains a self-instance, and passes along that self-instance. This means there should only be a single connection to the database, and allows the manager to work between two threads, necessary once the bot introduces a web UI.

On top of this, the DatabaseManager class can implement reference counting as part of its in-built context manager to determine when to finally close the connection.

As the DatabaseManager will no longer be passed around, there is a lot of code to update, with each place it is needed being wrapped in a `with` statement. To ease the pain of deeply nested indentation, I also propose having a generator function that yields the DatabaseManager inside a `with` statement. This allows a single function call that can handle cleanup without increasing the scope level anywhere a DB Manager is used.

### TL;DR and proposed goals:
- [ ] Have `DatabaseManager` statically wrap a self-instance to be used
- [ ] Implement reference counting to determine when to drop the DB connection
- [ ] Remove function parameters passing around a single instance of a `DatabaseManager`
- [ ] Implement a generator function to wrap `with` statements (to prevent nesting an extra scope)

**Comment by bastisawesome** [comment]
Began doing some testing on this, and a generator is not quite the correct solution. There is also some question about whether or not this is the correct thing to do, more on that below. For now, here is a basic mock example of how to wrap a context manager inside a function that returns the context-wrapped object:

```python
class TestClass:
    def __init__(self):
        pass

    def __enter__(self):
        return self

    def __exit__(_):
        pass

    def test_func(self):
        print("Test!")

    def __del__(self):
        print("TestClass deleted")

def test_generator():
    with TestClass() as t:
        return t


def test_the_class():
    t = test_generator()
    t.test_func()

test_the_class()
```

While I personally think this is a decent enough solution, it avoids needing to nest arbitrarily-deep, there are some potential reasons to avoid this:

1. It muddies the execution order significantly. In this example, `test_the_class` is called, which calls `test_generator`, which returns an instance of `TestClass`. However, most people would expect that after the line `t = test_generator()`, the function `test_generator` will be finished and have no further execution. However, because it wraps a context manager, `test_the_class` actually returns _first_, which then runs the cleanup code in `test_generator`. So the execution order can be confusing.
2. This is more of a retrospective on this entire concept: the entire idea that this encapsulates, a reference-count-controlled static class that controls a single instance of itself, is based on one major assumption: there will always be _at least one_ reference to `DatabaseManager`. While this constraint is provably true _in the current codebase_, the constraint could be made false in the future. This poses a new question: should the connection to the database always be maintained? There isn't a good answer for this; creating multiple connections throughout the lifetime of the program adds a lot of overhead, since every time the `DatabaseManager` is instanced, it also checks to ensure the database both exists and has been generated.

With all of this in mind, I still think this is a good idea, but I will sit on this for a while as I suspect this will be worked into 2.0.0 at the earliest and 2.1.0 if 2.0.0 ends up being a much larger update than originally intended (its biggest feature is simply the web UI, but there are a lot of smaller changes slated for that release).

Edit:
I was reviewing the current roadmap and this _could_ be slated into 1.1.0, although I feel like this is a big enough change that it should be part of a major revision. My personal belief is that anything that changes the existing ABI/API should mark a bump in the major version. Which would make this ineligible for 2.1.0 as well... Will review this during 2.0.0, as this would need to be done _before_ the web UI.
